### PR TITLE
Add client builder

### LINKED ISSUE #461: Use a factory or a configurator for the HTTP client
We should refactor away all the code in `Github\Client` into a HTTP client configurator of some kind. That would increase interoperability and clear up the responsibility.

### PR DESCRIPTION
**Pull request by Nyholm**
This will fix #461

The client builder makes sure that all logic related to creating the client is in a separate class. 
This is a first draft and feedback is welcome. 

### TODO

* [x] Fix the tests
* [x] Update docs

### CODE REVIEW

**Comment by acrobat** [comment]
I like the idea! :+1:

**Comment by acrobat** [comment]
@Nyholm you should also update the docs for [customizing the api-client](https://github.com/KnpLabs/php-github-api/blob/master/doc/customize.md) as `setHttpClient` is now removed

**Comment by Nyholm** [comment]
Thank you. I've updated the PR

**Comment by acrobat** [comment]
A few more places which use the old constructor

- https://github.com/KnpLabs/php-github-api/blob/master/doc/api_version.md#L11
- https://github.com/KnpLabs/php-github-api/blob/master/doc/security.md#L57

File: test/Github/Tests/HttpClient/BuilderTest.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+        $builder->expects($this->once())
+            ->method('removePlugin')
+            ->with(Plugin\HeaderAppendPlugin::class);
+
+        $builder->clearHeaders();
+    }
+
+
+
```
Comment (cursedcoder) [inline test/Github/Tests/HttpClient/BuilderTest.php]: too many empty lines?

File: lib/Github/HttpClient/Builder.php [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+use Http\Discovery\MessageFactoryDiscovery;
+use Http\Discovery\StreamFactoryDiscovery;
+use Http\Message\MessageFactory;
+use Http\Message\RequestFactory;
+use Http\Message\StreamFactory;
+use Psr\Cache\CacheItemPoolInterface;
+
+/**
+ * A builder that builds the API client. This will allow you to fluently add and remove
+ * plugins.
```
Comment (cursedcoder) [inline lib/Github/HttpClient/Builder.php]: you can place each sentence on new line, right now it feels a bit strange :D

File: lib/Github/Client.php [hunk 3]
```diff
@@ -146,27 +104,26 @@ class Client
...
      *
-     * @param HttpClient|null $httpClient
-     * @param string|null     $apiVersion
-     * @param string|null     $enterpriseUrl
+     * @param Builder|null $httpClient
+     * @param string|null  $apiVersion
+     * @param string|null  $enterpriseUrl
      */
-    public function __construct(HttpClient $httpClient = null, $apiVersion = null, $enterpriseUrl = null)
+    public function __construct(Builder $httpClientBuilder = null, $apiVersion = null, $enterpriseUrl = null)
```
Comment (cursedcoder) [inline lib/Github/Client.php]: IMO it's better to leave old constructor and create new builder inline or support both.

Comment (acrobat) [inline lib/Github/Client.php]: Maybe keep the old constructor indeed but deprecated it (throw deprecation warning) and use the new constructor in 3.0.

Comment (cursedcoder) [inline lib/Github/Client.php]: My idea is that client should always be constructed without extra manipulations, i.e. 

```
$github = new Client();
$github->doAnything();
```

maybe static factory can be a possible solution.

Comment (acrobat) [inline lib/Github/Client.php]: You are right about that, it should be a simple as possible to instantiate and use the client

Comment (Nyholm) [inline lib/Github/Client.php]: I did not like the new constructor myself to be honest. Should I add another constructor argument with the builder?

**Comment by Nyholm** [comment]
I made an update. Are we happy with this constructor?

File: doc/customize.md [hunk 4]
```diff
@@ -37,7 +36,10 @@ class CustomUserAgentPlugin implements Plugin
     }
 }
 
-$githubClient->addPlugin(new CustomUserAgentPlugin());
+$httpBuilder = new Github\HttpClient\Builder(new Http\Adapter\Guzzle6\Client());
+$httpBuilder->addPlugin(new CustomUserAgentPlugin());
+
+$client = new Github\Client(null, $httpBuilder);
```
Comment (Nyholm) [inline doc/customize.md]: Check out this example. 

Are we happy with the constructor like this?

Comment (acrobat) [inline doc/customize.md]: Hmm, I don't know. It looks a bit weird to me that you must start with `null`. I will think about a different approach

Comment (acrobat) [inline doc/customize.md]: That would maybe be a good solution. What do you think @cursedcoder?

Comment (cursedcoder) [inline doc/customize.md]: Seems nice.
Just `Client::fromAdapter()` or `fromClient`, put some more sense in your method name.

Comment (Nyholm) [inline doc/customize.md]: Great! I'll update the PR after lunch

Comment (Nyholm) [inline doc/customize.md]: `Client::withHttpClient($guzzle)`?
`Client::usingHttpClient($guzzle)`?
`Client::createWithHttpClient($guzzle)`?

Comment (acrobat) [inline doc/customize.md]: `Client::createFromHttpClient($guzzle)`? Otherwise I would go for `createWith..`

**Comment by Nyholm** [comment]
There, this PR is updated. I'm happy to change the name of `createFromHttpClient` to whatever. 

My *opinion* is that you would not create a Github client **from** a http client. The Github client is **using** the http client somehow.

**Comment by acrobat** [comment]
YEs maybe you are right about the name of the method!

**Comment by Nyholm** [comment]
> YEs maybe you are right about the name of the method!

I have no strong opinions at all. =) @cursedcoder can you make a decision and I can consider this PR ready for final review.

**Comment by cursedcoder** [comment]
> My opinion is that you would not create a Github client from a http client. The Github client is using the http client somehow.

Got you. So it makes sense to rename `createWithHttpClient`?

**Comment by Nyholm** [comment]
Thank you. I've updated the PR

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,20 @@ The change log describes what is "Added", "Removed", "Changed" or "Fixed" betwee
 
 ## 2.0.0-rc5 (UNRELEASED)
 
+### Changed
```
Comment (acrobat) [inline CHANGELOG.md]: The changed part is already added for RC5 right under added (line 32 in your changed file)

Comment (Nyholm) [inline CHANGELOG.md]: That's correct. Thank you.

**Comment by Nyholm** [comment]
Thank you for merging

**Pull request merged**